### ISSUE TITLE
ERROR: UndefVarError: last_f not defined (Julia v0.7.0-alpha)

### ISSUE THREAD

**Issue by bensadeghi** [bensadeghi/DecisionTree.jl#66]
There's a strange error when running ```build_tree``` in Julia v0.7.0-alpha (for both classification and regression), error traces below. What's peculiar is that the routines run fine in Julia v0.6.x.

For classification, I've tried a hack of adding ```last_f = Xf[1]``` before the ```while hi < n_samples``` loop (line 130 of tree.jl).
But with this change, I'm seeing some serious tree build performance degradation, both in execution time and memory allocation, on bleeding edge Julia. This performance issue might be due to something unrelated, but finding its root cause would be helpful for the Julia core team.

cc @Eight1911 

Classification:
```
ERROR: UndefVarError: last_f not defined
Stacktrace:
 [1] _split!(::Array{Float64,2}, ::Array{Int64,1}, ::DecisionTree.treeclassifier.NodeMeta, ::Int64, ::Int64, ::Int64, ::Int64, ::Int64, ::Float64, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Float64,1}, ::Array{Int64,1}, ::Random.MersenneTwister) at /home/bs/.julia/dev/DecisionTree/src/classification/tree.jl:131
 [2] #build_tree#1(::Any, ::Any, ::Array{Float64,2}, ::Array{Int64,1}, ::Int64, ::Int64, ::Int64, ::Int64, ::Float64) at /home/bs/.julia/dev/DecisionTree/src/classification/tree.jl:322
 [3] #build_tree at ./none:0 [inlined]
 [4] #build_tree#6 at /home/bs/.julia/dev/DecisionTree/src/classification/main.jl:86 [inlined]
 [5] build_tree at /home/bs/.julia/dev/DecisionTree/src/classification/main.jl:73 [inlined] (repeats 2 times)
 [6] top-level scope at none:0
```

Regression:
```
ERROR: UndefVarError: last_f not defined
Stacktrace:
 [1] _split!(::Array{Float64,2}, ::Array{Float64,1}, ::DecisionTree.treeregressor.NodeMeta, ::Int64, ::Int64, ::Int64, ::Int64, ::Float64, ::Array{Int64,1}, ::Array{Float64,1}, ::Array{Float64,1}, ::Random.MersenneTwister) at /home/bs/.julia/dev/DecisionTree/src/regression/tree.jl:139
 [2] #build_tree#1(::Random.MersenneTwister, ::Any, ::Array{Float64,2}, ::Array{Float64,1}, ::Int64, ::Int64, ::Int64, ::Int64, ::Float64) at /home/bs/.julia/dev/DecisionTree/src/regression/tree.jl:289
 [3] #build_tree at ./none:0 [inlined]
 [4] #build_tree#24 at /home/bs/.julia/dev/DecisionTree/src/regression/main.jl:126 [inlined]
 [5] build_tree at /home/bs/.julia/dev/DecisionTree/src/regression/main.jl:110 [inlined] (repeats 2 times)
 [6] top-level scope at none:0
```

**Comment by Eight1911** [comment]
This issue is due to the change in the scoping of `while` loops in 0.7. (compare [version 0.7](https://docs.julialang.org/en/latest/manual/variables-and-scoping/#For-Loops-and-Comprehensions-1) and [version 0.6](https://docs.julialang.org/en/v0.6.2/manual/variables-and-scoping/#For-Loops-and-Comprehensions-1)) Personally, I think the issue should be fixed after the release of the beta or the release-candidate for Julia 0.7 when things are more stable.

Moreover, it seems very possible that Julia 0.6 and Julia 0.7, being quite different, are not optimized by the same code, so it might be a good idea to make a separate branch for 0.7.

**Comment by bensadeghi** [comment]
Sounds good.
Let's hold a bit for the beta release, and get a separate branch going for Julia v0.7

**Comment by bensadeghi** [comment]
resolved with PR #71
